src/plugins/data_views/common/data_views/data_views.ts

```diff
@@ -570,30 +577,28 @@ export class DataViewsService {
    * @param displayErrors  - If set false, API consumer is responsible for displaying and handling errors.
    */
   refreshFields = async (dataView: DataView, displayErrors: boolean = true) => {
-    if (!displayErrors) {
-      return this.refreshFieldsFn(dataView);
-    }
-
     try {
       await this.refreshFieldsFn(dataView);
     } catch (err) {
-      if (err instanceof DataViewMissingIndices) {
-        this.onNotification(
-          { title: err.message, color: 'danger', iconType: 'alert' },
-          `refreshFields:${dataView.getIndexPattern()}`
+      if (displayErrors) {
```

Comment:
I'm confused about the work this change is doing. As best I can tell it swallows errors since the error is caught but not re-thrown which isn't desired.

Reply:
@mattkime The previous code was causing an unhandled promise error which was not allowing to switch from SQL to regular mode in UI (for a data view with missing index)

```
if (!displayErrors) {
  return this.refreshFieldsFn(dataView);
}
```

<img width="779" alt="Screenshot 2023-01-31 at 10 15 43" src="https://user-images.githubusercontent.com/1415710/215718947-f2d7d344-7049-49d1-b718-fc74cadc1be5.png">

Reply:
Can you handle the exception in the api consumer instead of changing the api?

Reply:
@mattkime Thanks for your suggestion! The problem is that the consumer is data views service itself
https://github.com/elastic/kibana/blob/4c386d9d0e1f602de7734cc997c1cecf247f550b/src/plugins/data_views/common/data_views/data_views.ts#L894

Would you suggest to add try-catch there depending on `displayErrors`?

Reply:
No, at least in theory the error will propagate up to the api consumer's call.

Reply:
@kertal This might be an argument against simply disabling all error toasts. It would mean the potential for unhandled errors that could cause other problems. We can't both prevent unhandled errors AND pass errors to API consumers.

Reply:
> No, at least in theory the error will propagate up to the api consumer's call.

@mattkime `dataView.get()` could still return a data view object for its consumer if we suppress exceptions for `refreshField`. If we change as you suggest then the consumer code would not be able to proceed which conflicts with the purpose of this PR.

> We can't both prevent unhandled errors AND pass errors to API consumers.

Agree with that. I'm not happy with the PR changes either. I can close it if that's our conclusion.

Reply:
Agree, there should be one way to handle it: to suppress the errors or to pass errors to the API consumers. This should also be clearly communicated to consumers, currently it seems it's a mixed handling of those cases?

Reply:
>If we change as you suggest then the consumer code would not be able to proceed

@jughosta Perhaps we should set up some time to discuss this. There's something I'm not understanding here.

Reply:
@kertal 

>Agree, there should be one way to handle it: to suppress the errors or to pass errors to the API consumers. This should also be clearly communicated to consumers, currently it seems it's a mixed handling of those cases?

I think our language might be conflating things. Currently, the api either displays errors or throws errors - and thrown errors need to be caught by api consumer code. 'Suppressing errors' could ambiguously mean either of those. 

First we need to remove the possibility of errors from our communication regarding errors to successfully deal with errors. 🥲

Reply:
I created https://github.com/mattkime/kibana/pull/new/129020-inline-missing-index-error-mk to illustrate what I see as the path forward, particularly https://github.com/elastic/kibana/pull/150246/files#diff-eeffc98614da75c1f8e6ce9b07802c11dce0ae41f80b9540784f0f44c2e86133R307 - Now I think I did this before understanding what @jughosta was saying. 

@jughosta - It took me a second reading to understand what you were looking to do - 

>[@mattkime] dataView.get() could still return a data view object for its consumer if we suppress exceptions for refreshField. If we change as you suggest then the consumer code would not be able to proceed which conflicts with the purpose of this PR.

Basically you're saying you want to be able to get a data view even if you can't get the field list. This isn't something thats currently supported by the API. A data view without a field list is of very limited use in most circumstances. Its an interesting idea. Current API consumers always assume the full field list is there - they don't have a notion of a data view with an unloaded field list. This is something I'd like to see changed but it would be a significant effort on the part of API consumers.